pkg/tfgen/generate_schema.go

```diff
@@ -97,39 +99,68 @@ func gatherSchemaNestedTypesForMember(member moduleMember) map[string]*schemaNes
 func (nt *schemaNestedTypes) gatherFromMember(member moduleMember) {
 	switch member := member.(type) {
 	case *resourceType:
-		nt.gatherFromProperties(member, member.name, "", member.inprops, true)
-		nt.gatherFromProperties(member, member.name, "", member.outprops, true)
+		nt.gatherFromProperties(member, member.name, member.inprops, true, false, true)
+		nt.gatherFromProperties(member, member.name, member.outprops, false, false, true)
 		if !member.IsProvider() {
-			nt.gatherFromProperties(member, member.name, "", member.statet.properties, true)
+			nt.gatherFromProperties(member, member.name, member.statet.properties, false, true, true)
 		}
 	case *resourceFunc:
-		nt.gatherFromProperties(member, member.name, "", member.args, true)
-		nt.gatherFromProperties(member, member.name, "", member.rets, true)
+		nt.gatherFromProperties(member, member.name, member.args, true, false, true)
+		nt.gatherFromProperties(member, member.name, member.rets, false, false, true)
 	case *variable:
-		nt.gatherFromPropertyType(member, member.name, "", "", member.typ, true)
+		nt.gatherFromPropertyType(member, member.name, "", member.typ, false, false, true)
 	}
 }
 
-func (nt *schemaNestedTypes) declareType(
-	declarer declarer, namePrefix, name, nameSuffix string, typ *propertyType, pyMapCase bool) string {
+func (nt *schemaNestedTypes) declareType(declarer declarer, namePrefix, name string, typ *propertyType,
+	isInput, isState, pyMapCase bool) string {
 
 	// Generate a name for this nested type.
-	baseName := namePrefix + strings.Title(name)
+	typeName := namePrefix + strings.Title(name)
 
 	// Override the nested type name, if necessary.
 	if typ.nestedType.Name().String() != "" {
-		baseName = typ.nestedType.Name().String()
+		typeName = typ.nestedType.Name().String()
 	}
 
-	typeName := baseName + strings.Title(nameSuffix)
-	typ.name = typeName
+	if existing, ok := nt.nameToType[typeName]; ok {
+		contract.Assert(existing.declarer == declarer)
+
+		if existing.typ.equals(typ) && pyMapCase == existing.pyMapCase {
```

Comment:
Can it happen that two types were equal in versions of a provider <= N but then a new version brings a property which makes them different. Will that cause the type name change and an unexpected breaking change?

Reply:
I guess I'm asking why don't we want to append those name suffixes all the time unconditionally?